### PR TITLE
Multi Account Mixed Observability Observability Accelerator

### PR DESCRIPTION
**Pull request by iamprakkie**
This PR adds **Multi Account Mixed Observability Observability Accelerator** pattern.

### Change log

1. Monitoring Environment now has an EKS cluster built using `ObservabilityBuilder` with Grafana Operator and External Secrets using `GrafanaOperatorConstruct`. Uses ArgoCDAddOn for GitOps.

1. Sample apps for prod accounts will be taken from https://github.com/iamprakkie/aws-observability-accelerator.git instead of https://github.com/aws-samples/eks-blueprints-workloads.git. Final change will be updated to aws-observability/aws-observability-accelerator.

1. Identity, datasources, dashboards for Grafana Operator are from https://github.com/iamprakkie/aws-observability-accelerator.git. Final change will be updated to aws-observability/aws-observability-accelerator. Grafana Dashboard URLs taken from cdk.json.

1. Documentation covers all steps required to deploy the pattern.

1. Trust entities for AMP role and CW role now uses WorkspaceIamRole as principal.

1. `github-token` and `github-ssh-key` are stored in AWS SecretsManager in `pipeline` and `monitoring` account. Rest all are moved to SSM Securestring parameter.

1. gitOwner will be picked up from SSM parameter `/cdk-accelerator/pipeline-git-info`.

1. Clean up script takes care of deleting stack resources and stack which are not covered with pattern destroy option.

1. Can be used for multi-account with multiple-regions too.

### Work in progress

1. Architecture diagram.

1. sample-apps and grafana-operator-app git URLs to point out to `aws-observability/aws-observability-accelerator`.

### CODE REVIEW

File: lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        const addOns: Array<blueprints.ClusterAddOn> = [
+            new blueprints.addons.XrayAdotAddOn(),
+            new blueprints.addons.ClusterAutoScalerAddOn(),
+            new blueprints.addons.SecretsStoreAddOn(),
+            new blueprints.addons.ExternalsSecretsAddOn(),
+            new blueprints.addons.GrafanaOperatorAddon({
+                createNamespace: true,
+            }),
+            new GrafanaOperatorSecretAddon(),
+            // grafanaOperatorArgoAddonConfig, // ArgoCD config for grafana-operator
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts:35]: Remove these comments.

File: lib/multi-acc-new-eks-mixed-observability-pattern/amg-iam-setup.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import { NestedStack, NestedStackProps } from 'aws-cdk-lib';
+import * as blueprints from '@aws-quickstart/eks-blueprints';
+import { Construct } from 'constructs';
+import * as cdk from 'aws-cdk-lib';
+import { PolicyStatement, Effect, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/amg-iam-setup.ts:5]: fix GH warnings.

File: lib/multi-acc-new-eks-mixed-observability-pattern/cloudwatch-monitoring.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            new blueprints.addons.XrayAdotAddOn(),
+            new blueprints.addons.ClusterAutoScalerAddOn(),
+            new blueprints.addons.SecretsStoreAddOn(),
+            new blueprints.addons.CloudWatchLogsAddon({ 
+                logGroupPrefix: `/aws/eks/${Id}`,
+                logRetentionDays: 30
+            }),     
+            /* already part of enableNativePatternAddOns 
+            new blueprints.addons.AwsLoadBalancerControllerAddOn(), // part of enableOpenSourcePatternAddOns
+            new blueprints.addons.CertManagerAddOn(), // part of enableOpenSourcePatternAddOns
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/cloudwatch-monitoring.ts:45]: Remove comments

File: lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import { Construct } from 'constructs';
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts:1]: Why is this name GrafanaOperator Construct?

File: lib/multi-acc-new-eks-mixed-observability-pattern/CreateIAMRoleNestedStack.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import { NestedStack, NestedStackProps } from 'aws-cdk-lib';
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/CreateIAMRoleNestedStack.ts:1]: All our filenames are all sentence cases with dash and not mixed cases.

File: lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import * as blueprints from '@aws-quickstart/eks-blueprints';
+import { ObservabilityBuilder } from '@aws-quickstart/eks-blueprints';
+import { GrafanaOperatorSecretAddon } from './grafanaoperatorsecretaddon';
+
+export default class GrafanaOperatorConstruct {
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts:6]: All our filenames are all sentence cases with dash and not mixed cases.

File: lib/multi-acc-new-eks-mixed-observability-pattern/getSSMSecureString.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import { SSM } from 'aws-sdk';
+
+// function to get SSM Parameterstor Securestring value
+export async function getSSMSecureString(parameterName: string, region: string): Promise<string> {
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/getSSMSecureString.ts:4]: Why is this required? Cant we use External secrets to read SSM?

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+        amgWorkspaceUrl = amgInfo.workspaceURL;
+        const amgWorkspaceIAMRoleARN = amgInfo.workspaceIAMRoleARN;
+
+        // Props for cross-account trust role in PROD1 account to trust AMG from MON account, inorder to access PROD1's AMP
+        ampAssumeRoleName = "AMPAccessForTrustedAMGRole";
+        const AMPAccessRoleStackProps: CreateIAMRoleNestedStackProps = {
+            roleName: ampAssumeRoleName!,
+            trustArn: amgWorkspaceIAMRoleARN!,
+            actions: [
+                "aps:ListWorkspaces",
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:104]: Please move this as a file ourside and render it. Check bluperints addons for examples

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+                "cloudwatch:GetMetricData",
+                "logs:DescribeLogGroups",
+                "logs:GetLogGroupFields",
+                "logs:StartQuery",
+                "logs:StopQuery",
+                "logs:GetQueryResults",
+                "logs:GetLogEvents",
+                "ec2:DescribeTags",
+                "ec2:DescribeInstances",
+                "ec2:DescribeRegions",
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:152]: Please move this as a file ourside and render it. Check bluperints addons for examples

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+        // creating constructs
+        const ampConstruct = new AmpMonitoringConstruct();
+        const blueprintAmp = ampConstruct.create(scope, context.prodEnv1.account, context.prodEnv1.region);
+        const blueprintCloudWatch = new CloudWatchMonitoringConstruct().create(scope, context.prodEnv2.account, context.prodEnv2.region, PROD2_ENV_ID);
+        const blueprintAmg = new GrafanaOperatorConstruct().create(scope, context.monitoringEnv.account, context.monitoringEnv.region);
+
+        // Argo configuration per environment
+        // CHANGE ME FINALLY
+        // const prodArgoAddonConfig = createArgoAddonConfig(
+        //     'https://github.com/aws-observability/aws-observability-accelerator.git',
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:185]: remove commented code.

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+            .application("npx ts-node bin/multi-acc-new-eks-mixed-observability.ts")
+            .name("multi-account-COA-pipeline")
+            .owner(gitOwner)
+            .codeBuildPolicies([ 
+                new iam.PolicyStatement({
+                    resources: ["*"],
+                    actions: [    
+                        "sts:AssumeRole",
+                        "secretsmanager:GetSecretValue",
+                        "secretsmanager:DescribeSecret",
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:232]: Please move this as a file ourside and render it. Check bluperints addons for examples

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+        GRAFANA_NSWRKLDS_DASH_URL: namespaceWorkloadsDashUrl,
+        GRAFANA_NODEEXP_DASH_URL: nodeExporterDashUrl,
+        GRAFANA_NODES_DASH_URL: nodesDashUrl,
+        GRAFANA_WORKLOADS_DASH_URL: workloadsDashUrl
+    };
+
+    // ArgoCDAddOnProps.values = {
+    //     server: {  // By default argocd-server is not publicaly exposed. uncomment this section, if you need to expose using ALB
+    //         service: {
+    //             type: 'LoadBalancer'
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:382]: remove comments

File: scripts/multi-acc-new-eks-mixed-observability-pattern/amg-preconfig.sh [hunk 13]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
```
Comment (elamaran11) [inline scripts/multi-acc-new-eks-mixed-observability-pattern/amg-preconfig.sh:1]: instead of `scripts` rename the folder as `helpers`

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@iamprakkie This is AWSome work. I love the depth of work, i can see the effort you have put. Appreciate this. I have some comments on the PR.

File: lib/multi-acc-new-eks-mixed-observability-pattern/getSSMSecureString.ts (hunk 7 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/getSSMSecureString.ts:4]: This is used to get SSM parameter 
/cdk-accelerator/cdk-context for CodePipeline
/cdk-accelerator/pipeline-git-info for CodePipeline source repo
/cdk-accelerator/amg-info which holds AMG info like IAMRoleARN that are required to create cross account IAM role 

As these parameter values are used outside of EKS cluster, I am using SDK to get SSM value instead of External secrets.

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts (hunk 8 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:104]: Here I am using create-iam-role.ts (a generic nestedstack class to create IAM role with required name, trustARN, actions and resources) to create 2 cross-account access roles. trustARN here is AMG's IAM role. Please correct me if I didn't understand this comment properly.

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts (hunk 9 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:152]: Here I am using create-iam-role.ts (a generic nestedstack class to create IAM role with required name, trustARN, actions and resources) to create 2 cross-account access roles. trustARN here is AMG's IAM role. Please correct me if I didn't understand this comment properly.

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts (hunk 10 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:185]: I've intentionally kept this comment to swap to aws-observability/aws-observability-accelerator from             'https://github.com/iamprakkie/aws-observability-accelerator.git', till PR on aws-observability-accelerator is approved. I will remove this comment and change repo URL which will also fix GH warnings.

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts (hunk 12 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:382]: I've kept this comment for future reference in case if argocd service need to be front-ended with ALB. Please let me know if you still prefer to remove this comment.

File: scripts/multi-acc-new-eks-mixed-observability-pattern/amg-preconfig.sh (hunk 13 above)
Comment (iamprakkie) [reply scripts/multi-acc-new-eks-mixed-observability-pattern/amg-preconfig.sh:1]: renamed

File: lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts (hunk 1 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts]: done

File: lib/multi-acc-new-eks-mixed-observability-pattern/cloudwatch-monitoring.ts (hunk 3 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/cloudwatch-monitoring.ts]: done

File: lib/multi-acc-new-eks-mixed-observability-pattern/CreateIAMRoleNestedStack.ts (hunk 5 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/CreateIAMRoleNestedStack.ts]: done

File: lib/multi-acc-new-eks-mixed-observability-pattern/amg-iam-setup.ts (hunk 2 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/amg-iam-setup.ts]: done

File: lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts (hunk 4 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts]: renamed to grafana-operator-index.ts. This creates EKS cluster with grafana-operator and external-secrets-operator

File: lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts (hunk 6 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/GrafanaOperatorConstruct.ts]: done

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts (hunk 8 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts]: done - moved outside

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts (hunk 9 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts]: done - moved outside

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts (hunk 11 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts]: done - moved outside

File: package.json [hunk 14]
```diff
@@ -24,8 +24,9 @@
   },
   "dependencies": {
     "@aws-quickstart/eks-blueprints": "^1.11.3",
-    "constructs": "^10.0.0",
     "aws-cdk": "2.91.0",
+    "aws-sdk": "^2.1455.0",
```
Comment (elamaran11) [inline package.json:28]: why is sdk needed?

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+const logger = blueprints.utils.logger;
+
+let ampAccount: string;
+let ampRegion: string;
+let cwAccount: string;
+let cwRegion: string;
+let monAccount: string;
+let monRegion: string;
+
+type repoTypeValues = 'public' | 'private';
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:25]: Why do we see GH Warnings still?

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+import { AmgIamSetupStack, AmgIamSetupStackProps } from './amg-iam-setup';
+import { getAMPAccessPolicyDocument } from './amp-access-policy';
+import { getCWAccessPolicyDocument } from './cw-access-policy';
+import { getCodeBuildPolicyDocument } from './codebuild-policy';
+import { CreateIAMRoleNestedStack, CreateIAMRoleNestedStackProps } from './create-iam-role';
+import { getSSMSecureString } from './get-ssm-securestring';
+
+const logger = blueprints.utils.logger;
+
+let ampAccount: string;
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:18]: remove these variable definits and add it while being used

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+
+    readonly pipelineRegion = process.env.COA_PIPELINE_REGION! || process.env.CDK_DEFAULT_REGION!;
+
+    async buildAsync(scope: Construct) {
+
+        // All Grafana Dashboard URLs from `cdk.json` if present
+        clusterDashUrl = utils.valueFromContext(scope, "cluster.dashboard.url", undefined);
+        kubeletDashUrl = utils.valueFromContext(scope, "kubelet.dashboard.url", undefined);
+        namespaceWorkloadsDashUrl = utils.valueFromContext(scope, "namespaceworkloads.dashboard.url", undefined);
+        nodeExporterDashUrl = utils.valueFromContext(scope, "nodeexporter.dashboard.url", undefined);
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:69]: you can say `const ...`

File: lib/multi-acc-new-eks-mixed-observability-pattern/grafana-operator-index.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
+import { Construct } from 'constructs';
```
Comment (elamaran11) [inline lib/multi-acc-new-eks-mixed-observability-pattern/grafana-operator-index.ts:1]: rename this file to something meaning ful like monitoring account or something

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@iamprakkie AWSome work. some minor feedback. Also mkdocs.yml file update and readMe update for list of patterns is missing.

File: package.json (hunk 14 above)
Comment (iamprakkie) [reply package.json:28]: for getting ssm securestring cdk-context using SDK

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts (hunk 16 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts:18]: I am declaring them upfront as they are used in class `PipelineMultiEnvMonitoring` and in function createGOArgoAddonConfig

File: lib/multi-acc-new-eks-mixed-observability-pattern/grafana-operator-index.ts (hunk 18 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/grafana-operator-index.ts]: done. renamed to central-monitoring.ts

File: lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts (hunk 17 above)
Comment (iamprakkie) [reply lib/multi-acc-new-eks-mixed-observability-pattern/pipeline.ts]: moved into interface

File: mkdocs.yml [hunk 19]
```diff
@@ -25,22 +25,23 @@ theme:
...
-      - OSS Nginx Mon : patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md
       - OSS Apiserver Mon : patterns/existing-eks-observability-accelerators/existing-eks-apiserver-observability.md
-      - Mixed : patterns/existing-eks-observability-accelerators/existing-eks-mixed-observability.md
-    - New Cluster: 
+      - OSS Nginx Mon : patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md
+    - New Cluster:
       - AWS Native : patterns/single-new-eks-observability-accelerators/single-new-eks-awsnative-observability.md
+      - Graviton OSS : patterns/single-new-eks-observability-accelerators/single-new-eks-graviton-opensource-observability.md
+      - Mixed : patterns/single-new-eks-observability-accelerators/single-new-eks-mixed-observability.md
+      - Multi-Cluster Multi-Region Mixed : patterns/multi-new-eks-observability-accelerators/multi-acc-new-eks-mixed-observability.md
```
Comment (elamaran11) [inline mkdocs.yml:38]: Remove the word Mixed and make it `Mon` ?

**Review by elamaran11 — CHANGES_REQUESTED** [review]

File: mkdocs.yml [hunk 20]
```diff
@@ -25,22 +25,23 @@ theme:
...
-      - OSS Nginx Mon : patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md
       - OSS Apiserver Mon : patterns/existing-eks-observability-accelerators/existing-eks-apiserver-observability.md
-      - Mixed : patterns/existing-eks-observability-accelerators/existing-eks-mixed-observability.md
-    - New Cluster: 
+      - OSS Nginx Mon : patterns/existing-eks-observability-accelerators/existing-eks-nginx-observability.md
+    - New Cluster:
       - AWS Native : patterns/single-new-eks-observability-accelerators/single-new-eks-awsnative-observability.md
+      - Graviton OSS : patterns/single-new-eks-observability-accelerators/single-new-eks-graviton-opensource-observability.md
+      - Mixed : patterns/single-new-eks-observability-accelerators/single-new-eks-mixed-observability.md
+      - Multi-Cluster Multi-Region Mixed : patterns/multi-new-eks-observability-accelerators/multi-acc-new-eks-mixed-observability-for-mkdocs.md
```
[the remark this answers is not in the document]
Comment (iamprakkie) [reply mkdocs.yml]: updated

**Review by elamaran11 — APPROVED** [review]
@iamprakkie Finally. Looks AWSome. This is outstanding work. So happy it got to Amazing shape. Thankyou for the hardwork. LGTM

**Review by elamaran11 — CHANGES_REQUESTED** [review]
Change to Mon is not done

**Review by elamaran11 — APPROVED** [review]
LGTM
